File: js/src/gilded_rose.js
class Item {
  constructor(name, sellIn, quality){
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

class Shop {
  constructor(items=[]){
    this.items = items;
  }

  doubleDegrade(item) {
    return item.quality - 2
  }

  checkNegative(item) {
    item.quality < 0 ? item.quality = 1 : item.quality
  }

  checkOver50(item) {
    item.quality > 50 ? item.quality = 50 : item.quality
  }

  conjuredUpdate(item) {
    item.quality -= 2
    debugger
  }

  standardUpdate(item) {
    item.sellIn < 0 ? item.quality -= 2 : item.quality -= 1
    this.reduceSellDate(item)
  }

  reduceSellDate(item) {
    item.sellIn -= 1
    debugger
  }

  checkNegativeAndOver50(item){
    this.checkNegative(item)
    this.checkOver50(item)
    debugger
  }

  determineItem(item) {
    switch (item.name) {
      case 'Aged Brie':
        agedBrieUpdater(item);
        this.reduceSellDate(item)
      case 'Sulfuras, Hand of Ragnaros':
        sulfurasUpdater(item);
      case 'Backstage passes to a TAFKAL80ETC concert':
        backStagePassUpdater(item);
        this.reduceSellDate(item)
      case 'Conjured':
        this.checkNegativeAndOver50(item);
        this.reduceSellDate(item)
        this.conjuredUpdate(item);
      default:
        this.checkNegativeAndOver50(item);
        this.standardUpdate(item);
    }
  }

  updateQuality() {
    for (var i = 0; i < this.items.length; i++) {
      this.determineItem(this.items[i])
    }
    return this.items
  }
}


function agedBrieUpdater(item){
  item.quality += 1
}

function sulfurasUpdater(item){
  item.sellIn = false
  item.quality = 80
}

function backStagePassUpdater(item) {
  if (item.sellIn < 5) {
    item.quality += 3
  } else if (item.sellIn < 10) {
    item.quality += 2
  } else if (item.sellIn == 0) {
    item.quality = 0
  } else {
    item.quality += 1
  }
}

File: ruby/gilded_rose.rb
class GildedRose

  def initialize(items)
    @items = items
  end

  def update_quality()
    @items.each do |item|
      determine_item(item)
    end
  end

  def standard_degrade(item)
    item.quality -= 1
  end

  def double_degrade(item)
    item.quality -= 2
  end

  def check_negative(item)
    item.quality < 0 ? item.quality = 0 : item.quality
  end

  def check_over_50(item)
    item.quality > 50 ? item.quality = 50 : item.quality
  end

  def conjured_update(item)
    double_degrade(item)
  end

  def standard_update(item)
    reduce_sell_date(item)
    item.sell_in < 0 ? double_degrade(item) : standard_degrade(item)
  end

  def reduce_sell_date(item)
    item.sell_in -= 1
  end

  def check_negative_and_over_50(item)
    check_over_50(item)
    check_negative(item)
  end

  def determine_item(item)
    case item.name
      when 'Aged Brie'
        reduce_sell_date(item)
        aged_brie_updater(item)
        check_negative_and_over_50(item)
      when 'Sulfuras, Hand of Ragnaros'
        sulfuras_updater(item)
      when 'Backstage passes to a TAFKAL80ETC concert'
        reduce_sell_date(item)
        back_stage_pass_updater(item)
      when 'Conjured Mana Cake'
        reduce_sell_date(item)
        conjured_update(item)
        check_negative_and_over_50(item)
      else
        standard_update(item)
        check_negative_and_over_50(item)
    end
  end

  def aged_brie_updater(item)
    item.sell_in < 0 ? item.quality += 2 : item.quality += 1
  end

  def sulfuras_updater(item)
    item.quality = 80
  end

  def back_stage_pass_updater(item)
    if item.sell_in < 0
      item.quality = 0
    elsif item.sell_in < 5
      item.quality += 3
    elsif item.sell_in < 10
      item.quality += 2
    else
      item.quality += 1
    end
  end

end


class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end
